[JSC] Implement `String#repeat` in C++

sosukesuzuki opened the pull request:
#### 2526a45e199d29c41a19e2726f7b1fe672197529
<pre>
[JSC] Implement `String#repeat` in C++
<a href="https://bugs.webkit.org/show_bug.cgi?id=304631">https://bugs.webkit.org/show_bug.cgi?id=304631</a>

Reviewed by Yusuke Suzuki.

This patch changes to implement `String#repeat` in C++.

Previously, `String#repeat` only allocated a sequential buffer without building a rope string
when the original string length was 1. This patch changes the behavior to allocate a sequential
buffer even when the length is not 1, as long as the resulting string is short enough.

                                               TipOfTree                  Patched

string-repeat-not-resolving-fixed            0.8174+-0.0664     ?      0.8587+-0.0355        ? might be 1.0505x slower
string-repeat-single-resolving               1.7876+-0.1402            1.6992+-0.0705          might be 1.0520x faster
string-repeat-small-resolving                1.1884+-0.0411     ^      0.7413+-0.0943        ^ definitely 1.6031x faster
string-repeat-resolving-no-inline            1.5253+-0.2846     ^      1.0961+-0.0468        ^ definitely 1.3916x faster
string-repeat-single-not-resolving           1.7091+-0.1271            1.5642+-0.0656          might be 1.0926x faster
string-repeat-small-not-resolving            1.1239+-0.0927     ?      1.1445+-0.0643        ? might be 1.0184x slower
string-repeat-not-resolving                  1.3078+-0.1511     ?      1.3150+-0.0577        ?
string-repeat-arith                          5.5090+-0.2520     !      6.5450+-0.0464        ! definitely 1.1881x slower
string-repeat-not-resolving-no-inline        1.3218+-0.0293     ?      1.3260+-0.0629        ?
string-repeat-resolving-fixed                1.9260+-0.3135     ^      0.9126+-0.0318        ^ definitely 2.1104x faster
string-repeat-resolving                      1.5575+-0.1956     ^      1.1582+-0.0874        ^ definitely 1.3448x faster
string-repeat-single-japanese                3.8928+-0.1288            3.8227+-0.0975          might be 1.0183x faster
string-repeat-single-emoji                   1.1680+-0.0333            1.1477+-0.0139          might be 1.0177x faster
string-repeat-single-ascii                   1.8143+-0.2555            1.7250+-0.1485          might be 1.0517x faster

Test: JSTests/stress/string-repeat-complex.js

* JSTests/stress/string-repeat-complex.js: Added.
(shouldBe):
(throw.new.Error):
(shouldThrow):
(shouldThrow.longStr.repeat):
(shouldBe.test):
* Source/JavaScriptCore/builtins/StringPrototype.js:
(linkTimeConstant.repeatSlowPath): Deleted.
(repeat): Deleted.
* Source/JavaScriptCore/runtime/StringPrototype.cpp:
(JSC::StringPrototype::finishCreation):
(JSC::repeatString):
(JSC::repeatRope):
(JSC::JSC_DEFINE_HOST_FUNCTION):

Canonical link: <a href="https://commits.webkit.org/304955@main">https://commits.webkit.org/304955@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b49c18e90263656ceb72c9571463f833b550f6b7

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/137015 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/9375 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/48302 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/144758 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/89988 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/10078 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/9502 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/104777 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/76444 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/139960 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/7399 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/122777 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/85613 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/7036 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/4736 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/5347 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/128976 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/116377 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/40949 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/147513 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/135501 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/9058 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/41518 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/113130 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/9076 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/161/builds/7612 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/113460 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28815 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/6959 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/119048 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/63348 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/9106 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/37102 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/168282 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/8829 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/72672 "Built successfully") | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/43912 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/9047 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/8899 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation reviewed (commented):
You need to remove old JS repeat implementation.

Constellation reviewed (changes requested):
Commented.

Constellation commented on Source/JavaScriptCore/runtime/StringPrototype.cpp:
```diff
@@ -1706,6 +1707,136 @@ JSC_DEFINE_HOST_FUNCTION(stringProtoFuncConcat, (JSGlobalObject* globalObject, C
+    unsigned copied = sourceLength;
+    while (copied < resultLength) {
+        unsigned copyLen = std::min(copied, resultLength - copied);
+        memcpySpan(buffer.subspan(copied, copyLen), buffer.subspan(0, copyLen));
```
`buffer.subspan(0, copyLen)` should be `buffer.first(copyLen)`.

Constellation reviewed (approved):
r=me

sosukesuzuki committed:
[JSC] Implement `String#repeat` in C++
https://bugs.webkit.org/show_bug.cgi?id=304631

Reviewed by Yusuke Suzuki.

This patch changes to implement `String#repeat` in C++.

Previously, `String#repeat` only allocated a sequential buffer without building a rope string
when the original string length was 1. This patch changes the behavior to allocate a sequential
buffer even when the length is not 1, as long as the resulting string is short enough.

                                               TipOfTree                  Patched

string-repeat-not-resolving-fixed            0.8174+-0.0664     ?      0.8587+-0.0355        ? might be 1.0505x slower
string-repeat-single-resolving               1.7876+-0.1402            1.6992+-0.0705          might be 1.0520x faster
string-repeat-small-resolving                1.1884+-0.0411     ^      0.7413+-0.0943        ^ definitely 1.6031x faster
string-repeat-resolving-no-inline            1.5253+-0.2846     ^      1.0961+-0.0468        ^ definitely 1.3916x faster
string-repeat-single-not-resolving           1.7091+-0.1271            1.5642+-0.0656          might be 1.0926x faster
string-repeat-small-not-resolving            1.1239+-0.0927     ?      1.1445+-0.0643        ? might be 1.0184x slower
string-repeat-not-resolving                  1.3078+-0.1511     ?      1.3150+-0.0577        ?
string-repeat-arith                          5.5090+-0.2520     !      6.5450+-0.0464        ! definitely 1.1881x slower
string-repeat-not-resolving-no-inline        1.3218+-0.0293     ?      1.3260+-0.0629        ?
string-repeat-resolving-fixed                1.9260+-0.3135     ^      0.9126+-0.0318        ^ definitely 2.1104x faster
string-repeat-resolving                      1.5575+-0.1956     ^      1.1582+-0.0874        ^ definitely 1.3448x faster
string-repeat-single-japanese                3.8928+-0.1288            3.8227+-0.0975          might be 1.0183x faster
string-repeat-single-emoji                   1.1680+-0.0333            1.1477+-0.0139          might be 1.0177x faster
string-repeat-single-ascii                   1.8143+-0.2555            1.7250+-0.1485          might be 1.0517x faster

Test: JSTests/stress/string-repeat-complex.js

* JSTests/stress/string-repeat-complex.js: Added.
(shouldBe):
(throw.new.Error):
(shouldThrow):
(shouldThrow.longStr.repeat):
(shouldBe.test):
* Source/JavaScriptCore/builtins/StringPrototype.js:
(linkTimeConstant.repeatSlowPath): Deleted.
(repeat): Deleted.
* Source/JavaScriptCore/runtime/StringPrototype.cpp:
(JSC::StringPrototype::finishCreation):
(JSC::repeatString):
(JSC::repeatRope):
(JSC::JSC_DEFINE_HOST_FUNCTION):

Canonical link: https://commits.webkit.org/304955@main
